File: turtle-soup-bot/src/main/kotlin/io/github/kapu/turtlesoup/bridge/SurrenderHandler.kt
package io.github.kapu.turtlesoup.bridge

import io.github.kapu.turtlesoup.models.SurrenderVote
import io.github.kapu.turtlesoup.service.GameService
import io.github.kapu.turtlesoup.service.SurrenderVoteService
import io.github.kapu.turtlesoup.utils.MessageKeys
import io.github.kapu.turtlesoup.utils.MessageProvider
import io.github.oshai.kotlinlogging.KotlinLogging

/** 포기/투표 핸들러 */
class SurrenderHandler(
    private val gameService: GameService,
    private val voteService: SurrenderVoteService,
    private val messageProvider: MessageProvider,
) {
    /**
     * 항복 합의 처리 (플레이어 수에 따라 자동 조건 변경)
     * - 1인: 즉시 항복
     * - 2인: 2명 동의 필요
     * - 3인+: 3명 동의 필요
     */
    suspend fun handleConsensus(
        chatId: String,
        userId: String,
    ): String {
        logger.info { "handle_surrender_consensus chat_id=$chatId user_id=$userId" }
        voteService.requireSession(chatId)

        val players = voteService.resolvePlayers(chatId)
        val activeVote = voteService.activeVote(chatId)

        return activeVote
            ?.let { vote -> activeVoteResponse(chatId, players, vote) }
            ?: newVoteResponse(chatId, userId, players)
    }

    /**
     * 포기 투표 동의
     */
    suspend fun handleAgree(
        chatId: String,
        userId: String,
    ): String {
        logger.info { "handle_agree chat_id=$chatId user_id=$userId" }
        voteService.requireSession(chatId)

        return when (val result = voteService.approve(chatId, userId)) {
            SurrenderVoteService.VoteApprovalResult.NotFound ->
                messageProvider.get(MessageKeys.VOTE_NOT_FOUND)

            SurrenderVoteService.VoteApprovalResult.NotEligible ->
                messageProvider.get(MessageKeys.VOTE_NOT_FOUND)

            SurrenderVoteService.VoteApprovalResult.AlreadyVoted ->
                messageProvider.get(MessageKeys.VOTE_ALREADY_VOTED)

            SurrenderVoteService.VoteApprovalResult.PersistenceFailure ->
                messageProvider.get(MessageKeys.ERROR_INTERNAL)

            is SurrenderVoteService.VoteApprovalResult.Progress ->
                inProgressMessage(result.vote)

            is SurrenderVoteService.VoteApprovalResult.Completed ->
                votePassedMessage(chatId, result.vote)
        }
    }

    private suspend fun activeVoteResponse(
        chatId: String,
        players: Set<String>,
        vote: SurrenderVote,
    ): String =
        if (players.size == 1) {
            logger.info { "single_player_surrender chat_id=$chatId" }
            voteService.clear(chatId)
            executeSurrender(chatId)
        } else {
            inProgressMessage(vote)
        }

    private suspend fun newVoteResponse(
        chatId: String,
        userId: String,
        players: Set<String>,
    ): String {
        return when (val result = voteService.startVote(chatId, userId, players)) {
            is SurrenderVoteService.VoteStartResult.Immediate -> {
                logger.info { "vote_immediate_pass chat_id=$chatId" }
                executeSurrender(chatId)
            }

            is SurrenderVoteService.VoteStartResult.Started ->
                messageProvider.get(
                    MessageKeys.VOTE_START,
                    "required" to result.vote.requiredApprovals().toString(),
                    "current" to result.vote.approvals.size.toString(),
                )
        }
    }

    private fun inProgressMessage(vote: SurrenderVote): String {
        val remain = vote.requiredApprovals() - vote.approvals.size
        return messageProvider.get(
            MessageKeys.VOTE_IN_PROGRESS,
            "current" to vote.approvals.size.toString(),
            "required" to vote.requiredApprovals().toString(),
            "remain" to remain.toString(),
        )
    }

    private suspend fun votePassedMessage(
        chatId: String,
        updated: SurrenderVote,
    ): String {
        logger.info {
            "vote_passed chat_id=$chatId " +
                "approvals=${updated.approvals.size}/${updated.requiredApprovals()}"
        }
        return messageProvider.get(MessageKeys.VOTE_PASSED) + "\n\n" + executeSurrender(chatId)
    }

    private suspend fun executeSurrender(chatId: String): String {
        val result = gameService.surrender(chatId)

        val hintBlock =
            if (result.hintsUsed.isNotEmpty()) {
                val header =
                    messageProvider.get(
                        MessageKeys.SURRENDER_HINT_BLOCK_HEADER,
                        "hintCount" to result.hintsUsed.size.toString(),
                    )
                val items =
                    result.hintsUsed.mapIndexed { index, hint ->
                        messageProvider.get(
                            MessageKeys.SURRENDER_HINT_ITEM,
                            "hintNumber" to (index + 1).toString(),
                            "content" to hint,
                        )
                    }.joinToString("\n")
                header + items
            } else {
                ""
            }

        return messageProvider.get(
            MessageKeys.SURRENDER_RESULT,
            "solution" to result.solution,
            "hintBlock" to hintBlock,
        )
    }

    companion object {
        private val logger = KotlinLogging.logger {}
    }
}

File: game-bot-go/internal/turtlesoup/mq/surrender_handler.go
package mq

import (
	"context"
	"fmt"
	"strings"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/messageprovider"
	tsmessages "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/messages"
	tsmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/model"
	tssvc "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/service"
)

// SurrenderHandler: 항복 투표 및 처리 로직을 담당하는 핸들러
type SurrenderHandler struct {
	gameService *tssvc.GameService
	voteService *tssvc.SurrenderVoteService
	msgProvider *messageprovider.Provider
}

// NewSurrenderHandler: 새로운 SurrenderHandler 인스턴스를 생성한다.
func NewSurrenderHandler(gameService *tssvc.GameService, voteService *tssvc.SurrenderVoteService, msgProvider *messageprovider.Provider) *SurrenderHandler {
	return &SurrenderHandler{
		gameService: gameService,
		voteService: voteService,
		msgProvider: msgProvider,
	}
}

// HandleConsensus: 항복 투표를 시작하거나 현재 진행 상황을 반환한다. (만장일치 시 즉시 항복 처리)
func (h *SurrenderHandler) HandleConsensus(ctx context.Context, chatID string, userID string) (string, error) {
	if err := h.voteService.RequireSession(ctx, chatID); err != nil {
		return "", fmt.Errorf("require session failed: %w", err)
	}

	players, err := h.voteService.ResolvePlayers(ctx, chatID)
	if err != nil {
		return "", fmt.Errorf("resolve players failed: %w", err)
	}

	active, err := h.voteService.ActiveVote(ctx, chatID)
	if err != nil {
		return "", fmt.Errorf("get active vote failed: %w", err)
	}

	if active != nil {
		if len(players) == 1 {
			_ = h.voteService.Clear(ctx, chatID)
			return h.executeSurrender(ctx, chatID)
		}
		return h.inProgressMessage(*active), nil
	}

	startResult, err := h.voteService.StartVote(ctx, chatID, userID, players)
	if err != nil {
		return "", fmt.Errorf("start vote failed: %w", err)
	}

	switch startResult.Type {
	case tssvc.VoteStartImmediate:
		return h.executeSurrender(ctx, chatID)
	case tssvc.VoteStartStarted:
		return h.msgProvider.Get(
			tsmessages.VoteStart,
			messageprovider.P("required", startResult.Vote.RequiredApprovals()),
			messageprovider.P("current", len(startResult.Vote.Approvals)),
		), nil
	default:
		return "", fmt.Errorf("unknown vote start result")
	}
}

// HandleAgree: 항복에 동의 투표를 하고 결과를 반환한다.
func (h *SurrenderHandler) HandleAgree(ctx context.Context, chatID string, userID string) (string, error) {
	if err := h.voteService.RequireSession(ctx, chatID); err != nil {
		return "", fmt.Errorf("require session failed: %w", err)
	}

	result, err := h.voteService.Approve(ctx, chatID, userID)
	if err != nil {
		return "", fmt.Errorf("approve vote failed: %w", err)
	}

	switch result.Type {
	case tssvc.VoteApprovalNotFound:
		return h.msgProvider.Get(tsmessages.VoteNotFound), nil
	case tssvc.VoteApprovalNotEligible:
		return h.msgProvider.Get(tsmessages.VoteNotFound), nil
	case tssvc.VoteApprovalAlreadyVoted:
		return h.msgProvider.Get(tsmessages.VoteAlreadyVoted), nil
	case tssvc.VoteApprovalPersistenceFailure:
		return h.msgProvider.Get(tsmessages.ErrorInternal), nil
	case tssvc.VoteApprovalProgress:
		if result.Vote == nil {
			return h.msgProvider.Get(tsmessages.ErrorInternal), nil
		}
		return h.inProgressMessage(*result.Vote), nil
	case tssvc.VoteApprovalCompleted:
		if result.Vote == nil {
			return h.msgProvider.Get(tsmessages.ErrorInternal), nil
		}
		body, err := h.executeSurrender(ctx, chatID)
		if err != nil {
			return "", fmt.Errorf("execute surrender failed: %w", err)
		}
		return h.msgProvider.Get(tsmessages.VotePassed) + "\n\n" + body, nil
	default:
		return h.msgProvider.Get(tsmessages.ErrorInternal), nil
	}
}

func (h *SurrenderHandler) inProgressMessage(vote tsmodel.SurrenderVote) string {
	remain := vote.RequiredApprovals() - len(vote.Approvals)
	return h.msgProvider.Get(
		tsmessages.VoteInProgress,
		messageprovider.P("current", len(vote.Approvals)),
		messageprovider.P("required", vote.RequiredApprovals()),
		messageprovider.P("remain", remain),
	)
}

func (h *SurrenderHandler) executeSurrender(ctx context.Context, chatID string) (string, error) {
	result, err := h.gameService.Surrender(ctx, chatID)
	if err != nil {
		return "", fmt.Errorf("surrender failed: %w", err)
	}

	hintBlock := ""
	if len(result.HintsUsed) > 0 {
		header := h.msgProvider.Get(tsmessages.SurrenderHintBlockHeader, messageprovider.P("hintCount", len(result.HintsUsed)))
		items := make([]string, 0, len(result.HintsUsed))
		for i, hint := range result.HintsUsed {
			items = append(items, h.msgProvider.Get(
				tsmessages.SurrenderHintItem,
				messageprovider.P("hintNumber", i+1),
				messageprovider.P("content", hint),
			))
		}
		hintBlock = header + strings.Join(items, "\n")
	}

	return h.msgProvider.Get(
		tsmessages.SurrenderResult,
		messageprovider.P("solution", result.Solution),
		messageprovider.P("hintBlock", hintBlock),
	), nil
}
